keybot/winbot.go

```diff
@@ -0,0 +1,164 @@
+	buildWindowsTest := buildWindows.Flag("test", "Whether build is for testing").Bool()
+	buildWindowsCientCommit := buildWindows.Flag("client-commit", "Build a specific client commit").String()
+	buildWindowsKbfsCommit := buildWindows.Flag("kbfs-commit", "Build a specific kbfs commit").String()
+	buildWindowsSkipCI := buildWindows.Flag("skip-ci", "Whether to skip CI").Bool()
+
+	cancel := app.Command("cancel", "Cancel current")
+
+	dumplogCmd := app.Command("dumplog", "Show the last log file")
+
+	logFileName := path.Join(os.TempDir(), "keybase.build.windows.log")
+
+	cmd, usage, cmdErr := cli.Parse(app, args, stringBuffer)
+	if usage != "" || cmdErr != nil {
+		return usage, cmdErr
+	}
+
+	if bot.Config().DryRun() {
+		return fmt.Sprintf("I would have run: `%#v`", cmd), nil
+	}
+
+	switch cmd {
+	case cancel.FullCommand():
+		mutex.Lock()
+		defer mutex.Unlock()
+		if buildProcess == nil {
+			return "No build running", nil
+		}
+		if err := buildProcess.Kill(); err != nil {
+			return "failed to cancel build", err
+		}
+
+	case buildWindows.FullCommand():
+		smokeTest := true
+		skipCI := *buildWindowsSkipCI
+		testBuild := *buildWindowsTest
+		// Don't smoke, wait for CI or promote test if custom build
+		if *buildWindowsCientCommit != "" || *buildWindowsKbfsCommit != "" {
+			smokeTest = false
+			skipCI = true
+			testBuild = true
+		}
+
+		updateChannel := "None"
+		if testBuild {
+			updateChannel = "Test"
+		} else if smokeTest {
+			updateChannel = "Smoke"
+			if !skipCI {
+				updateChannel = "SmokeCI"
+			}
+		}
+
+		// TODO: use SMOKE_TEST and TEST like other scripts
+		cmd := exec.Command(
+			"cmd", "/c",
+			path.Join(os.Getenv("GOPATH"), "src/github.com/keybase/client/packaging/windows/dorelease.cmd"),
+			">",
+			logFileName,
+			"2>&1")
+		cmd.Env = append(os.Environ(),
+			"ClientRevision="+*buildWindowsCientCommit,
+			"KbfsRevision="+*buildWindowsKbfsCommit,
+			"SKIP_CI="+boolToEnvString(skipCI),
+			"UpdateChannel="+updateChannel,
+			"SlackBot=1",
+		)
+		msg := fmt.Sprintf("I'm starting the job `windows build`. To cancel run `!%s cancel`", bot.Name())
+		bot.SendMessage(msg, channel)
+
+		go func() {
+			err := cmd.Start()
+			mutex.Lock()
+			buildProcess = cmd.Process
+			mutex.Unlock()
+			err = cmd.Wait()
+
+			bucketName := os.Getenv("BUCKET_NAME")
+			if bucketName == "" {
+				bucketName = "prerelease.keybase.io"
+			}
+			sendLogCmd := exec.Command(
+				path.Join(os.Getenv("GOPATH"), "src/github.com/keybase/release/release.exe"),
+				"save-log",
+				"--bucket-name="+bucketName,
+				"--path="+logFileName,
+			)
+			resultMsg := "Finished the job `windows build`"
+			if err != nil {
+				resultMsg = "Error in job `windows build`"
+				// Send a log snippet too
+				index := 0
+				logContents, err := ioutil.ReadFile(logFileName)
+				if err != nil {
+					bot.SendMessage("Error reading "+logFileName+": "+err.Error(), channel)
+				}
+				if len(logContents) > 500 {
```

Comment:
do you intent to get the last 500 bytes or the last 500 lines?  seems like you're getting the last 500 bytes...

Reply:
yeah I know it's bytes, thought that might be enough to see what went wrong... open to suggestions

Reply:
ok, just might be weird to get the middle of the line.  Since you're reading it all into a buffer anyways, it wouldn't be too different to split on newlines.  You can split what you have into smaller strings, or use [bufio.ReadLine](https://golang.org/pkg/bufio/#Reader.ReadLine)